### ISSUE TITLE
RangeSize issues

### ISSUE THREAD

**Issue by yisug** [biomodhub/biomod2#101]
Dear devs,

When using the BIOMOD_RangeSize() function I have been running into 3 problems.

Issue 1) for multiple different scenarios I noticed a lot of data gets written to my Temp folder (C:/user/name/AppData/Local/Temp). While running 1 or 2 different scenarios everything works as expected. When I start running multiple scenarios (4 different RangeSize comparisons for 10 species), my drive will eventually fill up with 200+gB worth of data in my temp folder. The problem I then run into is that this data doesn't get removed when closing R or rebooting the computer. I can easily manage it by deleting unneeded scenarios after getting any data/plots I need, but it would be nice to have the RangeSize outputs saved in a predetermined path so they can be loaded at a later time/on a different machine.

Issue 2) When plotting different scenario's with the RangeSize function my plots all miss world map outlines (regardless of the extent of my used rasters). This is especially inconvenient for species with a small projected distribution as it is impossible to know where the projected distribution is located without the visual aid of having a map. I checked another similar issue (https://github.com/biomodhub/biomod2/issues/72) and I tried using the on_0_1000 argument in my projections. The missing outlines persisted regardless if I set the argument to TRUE or FALSE. I hope you can help me with this problem because my Biomod experience so far has been great and I'd like to continue using it!

Issue 3) I run into (both on my desktop and on my work laptop) is that my machines crash occasionally when running the RangSize function. After these crashes, I checked the event log in Windows. On both devices every crash is logged as follows:

Log Name: System
Source: Kernel-Power
Event ID: 41
Level: Critical
User: SYSTEM
Task Category: (63)

I updated my drivers on both devices but the problem still occurs once in every 3-4 runs of the function.

Thank you for your time!

EDIT: here is an example plot using the tutorial with GuloGulo data.
![image](https://user-images.githubusercontent.com/47861627/190877912-fbfc6ff5-c355-428c-bfba-e94fa9388f2e.png)

gg = bm_PlotRangeSize(bm.range = RangeSize1, 
                      do.count = TRUE,
                      do.perc = TRUE,
                      do.maps = TRUE,
                      do.mean = TRUE,
                      do.plot = TRUE,
                      row.names = "Species")

**Comment by rpatin** [comment]
Hello @yisug,
Thank you for reporting !
1. So concerning your first issue, we will try to work at a better management of temporary files but cannot do promises. Although saving of range size projection is not automatically done, you can still save manually your projection to use them later:
```
myBiomodRangeSize <-
  BIOMOD_RangeSize(
    proj.current = CurrentProj,
    proj.future = FutureProj
  )
# only the raster:
writeRaster(myBiomodRangeSize$Diff.By.Pixel, filename = "myRangeSize.grd", datatype = "INT2S", NAflag = -9999)
# the whole object (not sure this is efficient in terms of storage though):
saveRDS(myBiomodRangeSize, file = "myRangeSize.rds")
```

2. Concerning your second issue, the different plots were not designed to show world contours. They were represented on the previous graphics only because absences were worldwide. And absences that are stable between both current and future projection are removed from the plot with `bm_PlotRangeSize`, hence this is why you cannot see the world contour on those maps. We might change this behaviour in the future or add an option to see stable absences. Meanwhile you can:
    a. use the raw raster maps available in  `myBiomodRangeSize$Diff.By.Pixel`
    b. use coastline data available from external sources (e.g. the `rworldmap` package) to consistently show world contour. You may need to manually redo the graphics though. A quick example below:

```
library(ggplot2)
library(rworldmap)

gg = bm_PlotRangeSize(
  bm.range = myBiomodRangeSize, 
  do.count = TRUE,
  do.perc = TRUE,
  do.maps = TRUE,
  do.mean = TRUE,
  do.plot = TRUE,
  row.names = c("Algo")
)

data(coastsCoarse)
df_coastsCoarse <- fortify(coastsCoarse) #transform into a data.frame for ggplot2
ggplot(gg$tab.ca1)+ # redo the graphics
  geom_raster(aes(x=x,y=y, fill = factor(value)))+
  scale_fill_manual("", values = c("-2" = "#fc8d62"
                                   , "-1" = "grey"
                                   , "0" = "white"
                                   , "1" = "#66c2a5")
                    , labels = c("-2" = "Loss"
                                 , "-1" = "Stable1"
                                 , "0" = ""
                                 , "1" = "Gain")) +
  geom_path(data = df_coastsCoarse, aes(x = long, y = lat, group = group) ) # add costalines
``` 

3. I could not reproduce the issue so we may require a bit more information. The event log you provided is likely linked to when you shutdown your PC. I suppose during the crash the PC freezes and you finally press the power button until in shutdown (at least that's what the log seems to be telling).
If there is indeed a freeze as I suppose, it might be linked to a saturated RAM (which you could check by opening your Task Manager) or because you ran out of disk space (maybe due to the issue with temporary files you described above). Maybe you could investigate those two hypotheses ?

**Comment by yisug** [comment]
Thank you for the solution to the map outlines, my plots are good now!

So far I have done the rangesize 10 times for the tutorial, no crashes yet. I think it has something to do with my raster size. I'm using variable data in 30arcsec and the RangeSize() function takes over 30 minutes to finish computing. Maybe I need more RAM to continue my work at this resolution. What I can see when looking at my taskmanager is that RAM keeps filling up every time I'm using the function. After multiple runs the RAM usage doesn't decrease. Only after a force quit of Rstudio my RAM usage goes back to normal.

**Comment by rpatin** [comment]
Indeed, I guess that memory may saturate and cause crash with such big raster. 

Concerning the increasing RAM usage, it is a generic problem with Rstudio when you successively manipulate large objects in loops for which there are generally no miraculous solutions. You can however test:
1. try to remove objects in your workspace once you are done with a species, using:
`rm(list = c("object1","object2", "object3")`
2. after object removal or at the end of each iteration, call the garbage collector with `gc()`. Unfortunately it does not work perfectly and Rstudio accumulates memory leaks that can be cleaned only when restarting it (from my own experience at least)
Hope this may help.

**Comment by yisug** [comment]
Thanks for all your help and the explanation!

I tried removing the objects and that doesn't work for my RAM. Somehow when using RangeSize() Rstudio keeps my RAM filled. Instead of doing all rangesize executions in one go I will do them one by one and restart Rstudio after getting the necessary plots. Hopefully, that will prevent more crashes
